WT-6866 Refactor python backup tests initial base class

boboeee opened the pull request:
This ticket aims to refactor most of the python backup tests. There is currently alot of re-used functions are created locally in each backup test file. Therefore by creating a base class, we allow cleaner code, and remove the redundancy of functions in the tests. Currently an initial base class is defined, moving over the **setup_directories** and the **compare_backups** functions first, as a measure of how the base class will be formed and used across different backup tests

boboeee commented:
@sueloverso hi sue! this PR is ready for review now

sueloverso commented on test/suite/test_backup01.py:
```diff
@@ -31,15 +31,15 @@
 import time
-from suite_subprocess import suite_subprocess
-import wiredtiger, wttest
+from test_backup_base import test_backup_base
```
In the `test/suite` directory, only actual tests have a `test_` prefix to the filename. I think this should be in a file perhaps called `wtbackup.py` and functions named appropriately. I suggest looking at `wtdataset.py` for one example.

boboeee replied:
awesome, changed it to wtbackup.py

sueloverso commented on test/suite/test_backup14.py:
```diff
@@ -273,7 +230,8 @@ def add_data_validate_backups(self):
         self.take_incr_backup()
-        self.compare_backups(self.uri)
+        paths = self.setup_compare_backup_paths()
+        self.compare_backups(self.uri, paths['full_backup_dir'], paths['full_backup_out'], True, paths['incr_backup_dir'], paths['incr_backup_out'])
```
This one and all the others below are very complicated and not an improvement over the original. Is there some way to add a wrapper that just takes `paths` and pulls out those pieces in the wrapper? They're all identical as far as I see.

boboeee replied:
yeah the redundancy, caused Tammy and I to re-purpose how compare_backups should look like, would love your insight on the new method, feels alot cleaner

sueloverso commented on test/suite/test_backup_base.py:
```diff
@@ -0,0 +1,77 @@
+    # Compare against two directory paths using the wt dump command. If the second directory path is not provided
+    # it checks if the compare
+    #
+    def compare_backups(self, uri, dir1, file1, run_recovery = False, dir2 = None, file2 = None):
```
Why do we need output file names as args? As far as I can tell the tests don't use them anywhere else and don't care about them. Only this function uses them to compare the two directories so I don't think they need to be args and the function can just make its own file names.

boboeee replied:
the outfiles are useful when you need to perserve the files if you wanted to look at the test directory for further information. Since the outfiles are created in the base test directory, test files like test_backup14.py, test_backup15.py create different outfiles, such that you can visualise the different outfiles that are made by the test. If we wanna remove the output arguments, we can still maintain the backups by placing them into their individual directories instead of the base directory. What opinions do you have?

boboeee replied:
Update: talked with tammy, removed the outfiles arguments, with new version of **compare_backups**, now we pass in a "suffix" to account for the different possible outfiles

sueloverso reviewed (changes requested):
This is definitely going in the right direction. I have a few comments for changes.

boboeee commented on test/suite/test_backup_base.py:
```diff
@@ -0,0 +1,77 @@
+            if os.path.exists(file2):
+                os.remove(file2)
+            if run_recovery:
+                self.runWt(['-R', 'dump', uri], outfilename=file2)
```
Hey @sueloverso , wanted your insight on this, can we force recovery with **-R** with tests that are don't test the incremental backup? If we can, we wouldn't need the **run_recovery**

sueloverso replied:
This appears to be resolved.

boboeee replied:
Hey @sueloverso ! actually we still need your insight in this, we are still not 100% sure that we should forcing recovery on tests that don't need to recover at all

sueloverso replied:
My guess is that most uses simply started by copying an existing test and then if they passed, probably changes were not considered. I'd suggest trying without recovery and then re-evaluate if the test fails.

boboeee replied:
awesome, tested and the tests don't fail

sueloverso commented on test/suite/test_backup01.py:
```diff
@@ -31,15 +31,15 @@
 import time
-from suite_subprocess import suite_subprocess
-import wiredtiger, wttest
+from wtbackup import test_backup_base
```
The internal class name should also change name from `test_backup_base`. No only for cleanliness, but also the test harness library we use for python attributes special meaning to things (files and functions) prepended with `test_` and while it may work now, it could lead to subtle bugs if that library changes.

boboeee replied:
fair, will call it **backup_base**

sueloverso commented on test/suite/test_backup14.py:
```diff
@@ -330,23 +286,23 @@ def insert_bulk_data(self):
-        self.compare_backups(self.uri_logged)
-
+        self.compare_backups(self.uri_logged, self.home_full, self.home_incr, str(self.counter))
+        self.compare_backups(self.uri, self.home_full, self.home_incr, str(self.counter))
```
This line here is the only one that is a functionality change. You modified the line that compared `uri_logged` and then added this new comparison. Was that intentional? It was unexpected to me.

boboeee replied:
oops sorry, accidently added this line. Thanks!

sueloverso commented on test/suite/test_backup15.py:
```diff
@@ -293,7 +258,8 @@ def test_backup15(self):
                 self.take_incr_backup()
                 self.take_full_backup()
-            self.compare_backups(self.uri)
+
```
I find that the `compare` call next to the `take_backup` calls is clear and shows it is related to taking the backups. I would suggest you don't add this blank line.

boboeee replied:
yep! removed.

sueloverso commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,78 @@
+        # Run wt dump on incremental backup
+        other_dir = other_dir_home + sfx
+        self.runWt(['-R', '-h', other_dir, 'dump', uri], outfilename=other_out)
+        self.assertEqual(True, compare_files(self, base_out, other_out))
```
I'm seeing a strange indicator here that there is no newline at the end of the file.

boboeee replied:
done!

sueloverso reviewed (commented):
This is very close. I have a few comments and a question on a functionality change.

boboeee commented:
Hey sue! the plan for this PR is to add incremental changes and review them on a change by change review. So after this looks good, ill put in the next changes.

sueloverso reviewed (approved):
This set of changes LGTM.

boboeee commented:
Hey @sueloverso ! Added support for **add_data** in the base class now. it is ready for review now, let me know what you think especially the usage of options idea.

boboeee commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,108 @@
+    #   mult: counter to have variance in data
+    #   counter: a counter to used to produce unique backup ids
+    #   session_checkpoint: boolean in which enables session checkpoints after adding data
+    #   initial_backup: To determine whether to increase/decrease counter, which determines
```
I feel like we can remove **initial_backup** from here, and the backup tests that use this variable. Instead an option would be to increase the counter if the test needs it to, by assigning to the return value of the **add_data** function. @sueloverso  what are your thoughts?

sueloverso replied:
@ddanderson can you take a look at this PR and the use of `data_options`? Do you have any suggestions on the best way to refactor and use these arguments? Does it overlap with the `scenarios` or other underlying Python functions we already have?

ddanderson commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,108 @@
+    #   initial_backup: To determine whether to increase/decrease counter, which determines
+    #           the backup id, and is generally used only for using add_data() first time.
+    #
+    def add_data(self, uri, key, val, options):
+        c = self.session.open_cursor(uri, None, options.get('cursor_options'))
```
I think using an options "dictionary" to pass around values, feels a bit cumbersome.  Rather, I would have the shared base class create the variables with their defaults:
```
class backup_base(....):
   nops = 100
   mult = 0
   # etc....
```
And the other classes can override them:
```
class test_backup16(backup_base):
   mult=1
   nops=10
   # etc....
```
Or the scenarios can override the values if needed.  (I'm not sure we do that now, but we could have a scenario with `nops=1000` in the dict.  Then, we should be able to use `self.counter`, `self.mult`, `self.nops` directly.  For example:
```
self.data_options['counter'] += 1
```
can be like the original:
```
self.counter += 1
```
I think that will be easier to understand and maintain.  And the number of changes needed to be made should be reduced.

boboeee replied:
sounds good, i've made the changes, only backup17 required an special change, let me know how you think of that test :D

ddanderson reviewed (commented):
Hello @boboeee, I have some comments about a different approach that I hope would result in code that is easier to understand and maintain.  If you disagree, or see some problems with the approach, let's talk about it.  Thanks!

boboeee commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,108 @@
+    # Compare against two directory paths using the wt dump command. The suffix allows the option
+    # of create output files to distinct.
+    #
+    def compare_backups(self, uri, base_dir_home, other_dir_home, suffix = None):
```
True! i'll update the comment

ddanderson reviewed (approved):
This looks great, so easy to review.  I made a minor suggestion, take it if it makes sense, or leave it.

boboeee commented on test/suite/test_backup17.py:
```diff
@@ -121,9 +111,10 @@ def test_backup17(self):
-        self.mult += 1
+        self.add_data(self.uri, self.bigkey, self.bigval, True)
+        self.mult = 0
+        self.add_data(self.uri2, self.bigkey, self.bigval, True)
+        self.mult = 1
```
done!

boboeee commented on test/suite/test_backup17.py:
```diff
@@ -150,10 +141,12 @@ def test_backup17(self):
         # verify we get fewer, consolidated values.
-        self.add_data(self.uri)
+        self.add_data(self.uri, self.bigkey, self.bigval, True)
+        self.mult = 1
```
done!

boboeee commented on test/suite/test_backup17.py:
```diff
@@ -150,10 +141,12 @@ def test_backup17(self):
 
-        self.add_data(self.uri2)
+        self.add_data(self.uri2, self.bigkey, self.bigval, True)
+        self.mult = 1
```
done!

boboeee commented:
Hey @ddanderson  and @sueloverso ! I've added in the next set of changes, these changes should be quite small. I had some git problems, so the commits might look abit over the place. I have added in **populate** and **confirmPathDoesNotExist** function in wtbackup.py. The files that were affected were mainly test_backup01.py, test_backup03.py and test_backup04.py. I'm not fond of the populate function so let me know what you guys think! 👍

ddanderson commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,147 @@
+            self.counter += 1
+        # Increase the multiplier so that later calls insert unique items.
+        self.mult += 1
+    
```
There's a minor thing that I think you'll see if you do this in your tree:
```
$ git diff develop test/suite/wtbackup.py
```
There are a few places where extra spaces at the end of lines are shown highlighted in that file.  Also you can detect by:
```
$ grep -n ' $' test/suite/wtbackup.py 
```
I wouldn't care too much, except that `git diff` seems to care, so the diffs don't look very good.

sueloverso replied:
It appears that this is also the case in `test_backup17.py`. Whitespace in `test_backup17.py` and `wtbackup.py` are causing the PR testing to fail. So this does need to be fixed.

boboeee replied:
Sorry forgot to run s_all! i think it fixes s_all

ddanderson reviewed (approved):
LGTM.  Noticed a tiny issue, no need to re-review.

sueloverso commented on test/suite/test_backup14.py:
```diff
@@ -26,64 +26,37 @@
-                os.remove(remove_dir)
-            os.makedirs(create_dir)
-
+    nops = 1000
```
I'm not sure why this was moved to here from the group of initializations above.

boboeee replied:
think it was a bad merge on my part, ill move it back 👍

sueloverso commented on test/suite/wtbackup.py:
```diff
@@ -0,0 +1,147 @@
+    nops = 100                  # number of operations added to uri
+
+    # counter to used to produce unique backup ids the backup id, and is
+    # generally used only for using add_data() first time.
```
This comment is difficult to parse and understand. How about something like:
We use counter to produce unique backup names for multiple iterations of incremental backup tests.

boboeee replied:
thanks for catching it! think i made a mistake in the commenting for counter

tammybailey reviewed (approved):
lgtm
